docs/developer/decision-records/2023-03-12_testing_pipeline/README.md

```diff
@@ -0,0 +1,124 @@
+  the `vault` service
+- `End-To-End-Test`: runs all tests tagged with `@EndToEndTest` using Minio
+- `Component-Test`: runs all tests tagged with `@ComponentTest`
+- `Api-Test`: runs all tests tagged with `@ApiTest`
+
+With this structure, there are some problems, which are:
+
+- Tests are run in wrong GH Actions jobs, e.g. `End-To-End` test only executes the inmem variant, the Postgres e2e
+  variant is run in the `Postgres-Integration-Test` job
+- Tests are incorrectly categorized:
+    - The concrete `EndToEnd___Test` instances are tagged with `@PostgresqlIntegrationTest`
+      or `@AzureCosmosIntegrationTEst`, but not with `@EndToEndTest`
+    - End-to-end tests tagged with `@PostgresqlIntegrationTest` are executed during the `Postgresql-Integration-Test`
+      job, and not during the End2End job.
+    - The `EndToEndTransferCosmosDbTest` is never executed due to incorrect/missing tags
+- Conditional execution based on the existence of secrets (i.e. CosmosDB) is cumbersome (fixed in my latest PR)
+
+## Approach / Solution proposals
+
+### Avoid running untagged tests
+
+Whenever a tag inclusion is specified (`-DincludeTags="..."`), we should _only_ run tests tagged as such. We must avoid
+running untagged tests to reduce test runtime, and more clearly delineate test runs.
+For that, a small modification to the `TestConvention` in the plugins repo is necessary.
+
+### Create dedicated Gradle tasks
+
+In order to avoid having to specify the exact tag name
+as `./gradlew test -DincludeTags="PostgresqlIntegrationTest,AzureCosmosDbIntegrationTest"` we should create dedicated
+Gradle tasks to make this more robust toward refactoring. The same command could then be written as:
+
+```bash
+./gradlew clean testPostgres testAzureCosmos
+```
+
+For example, we could do the following:
+
+```java
+
+@Tag("Api")
+public class SomeApiTest {
+    @Test
+    void foo() {
+        fail("api is broken!");
+    }
+}
+```
+
+and correspondingly in the `build.gradle.kts`:
+
+```kotlin
+tasks.register<Test>("apiTest") {
+    useJUnitPlatform {
+        includeTags("Api")
+    }
+    group = "verification"
+}
+```
+
+Then, we can execute `./gradlew apiTest`, which will only run the test tagged with `@Tag("Api")`.
+
+A small Gradle Plugin creating these tasks could be implemented in the `buildSrc` directory.
+
+**We must adapt our build plugin's `TestConvention`, because currently it would exclude all `@IntegrationTest` tests
+unless the
+`includeTags` system property is specified!**
+
+### Refactoring the CI Pipeline
+
+We propose to refactor the current pipeline from its currently very linear form with a single workflow into several
+workflows:
+
+- `Verify.yaml`: all other jobs depend on this one
+    - `Checkstyle`: job that runs checkstyle
+    - `Dependency-Analysis`: runs the dependency analysis
+- `Unit-Test.yaml`: runs unit tests with coverage and uploads test reports
```

Comment:
If I remember correctly unit & integration tests were put in the same workflow to upload coverage and test reports in the same run, if we separate them I guess we'll need to define a dedicated workflow that will run only when the dependent ones are finished

Reply:
To be honest, I'm not sure if we really need those coverage reports and test reports. Is anyone every looking at them? AFAIK there is a novel way to incorporate Codecov, maybe that's all we need?

Reply:
ok, so what we'll probably want is an aggregated Jacoco report across all modules, right? Currently, we only have separate reports _per module_. 
That aggregate report can then be uploaded to Codecov.

The problem with that is though, since we don't run all the tests (and thus coverage repots) in one job or even workflow, we never have all reports in one place. Maybe we can upload all reports separately as GH artifacts and publish them to Codecov in a separate action?

Reply:
I'm not really a codeconv expert, but if that's possible, it could be the way to go

Reply:
neither am I, but AFAIK Codecov accepts jacoco reports, and it probably doesn't care whether it's a specific one or the aggregated one. 
outputting them as artifacts, then downloading and aggregating them will likely be greater challenge.